File: frontend/types/listing.ts
/**
 * TypeScript type definitions for the Listings API.
 * Accurately mirrors the backend Pydantic models (HostResponse, ListingResponse).
 */

export interface Host {
  id?: number;
  name: string;
}

export interface Listing {
  id: number;
  title: string;
  description: string;
  city: string;
  country: string;
  latitude: number | null;
  longitude: number | null;
  price_per_night: number;
  cleaning_fee: number;
  max_guests: number;
  rating_avg: number | null;
  review_count: number;
  host: Host;
  host_id?: number;
  host_name?: string;
  images: string[];
  amenities: string[];
}

export interface ListingAvailability {
  listing_id: number;
  unavailable_dates: string[];
  booked_ranges: Array<{ check_in: string; check_out: string }>;
}

/**
 * Paginated response envelope returned by GET /api/listings.
 * Mirrors backend ListingsPageResponse.
 */
export interface ListingsPage {
  items: Listing[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

/**
 * Conceptual type representing an image entity.
 * Note: The API returns listing images as an array of URL strings (images: string[]).
 */
export interface ListingImage {
  id?: number;
  url: string;
  caption?: string | null;
  position?: number;
}

/**
 * Conceptual type representing an amenity entity.
 * Note: The API returns amenities as an array of name strings (amenities: string[]).
 */
export interface Amenity {
  id?: number;
  name: string;
}

File: backend/app/schemas/listing.py
"""
Pydantic response models for listing API endpoints.

These define the API contract — what the client receives.
They do NOT mirror the database schema directly.
"""

from pydantic import BaseModel


class HostResponse(BaseModel):
    """Public host information."""
    id: int | None = None
    name: str


class ListingResponse(BaseModel):
    """Full listing representation returned by the API."""
    id: int
    title: str
    description: str
    city: str
    country: str
    latitude: float | None
    longitude: float | None
    price_per_night: float
    cleaning_fee: float
    max_guests: int
    rating_avg: float | None
    review_count: int
    host: HostResponse
    host_id: int | None = None
    host_name: str | None = None
    images: list[str]
    amenities: list[str]


class ListingsPageResponse(BaseModel):
    """Paginated response envelope for the listings collection endpoint."""
    items: list[ListingResponse]
    page: int
    page_size: int
    total: int
    total_pages: int
